### ISSUE TITLE
Language model inference

### ISSUE THREAD

**Issue by gonzalobenegas** [HazyResearch/hyena-dna#20]
Thank you for your great work!

While the Huggingface examples seem to be about embeddings,  I would love to do inference with the language modeling head. I'm particularly interested in doing variant effect prediction using the log-likelihood of a sequence (as in [this protein paper](https://www.nature.com/articles/s41467-021-22732-w)). It would also be helpful if you could implement the [AutoModelForCausalLM API](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForCausalLM).

**Comment by exnx** [comment]
Thank you!  Much appreciated!

That's an interesting use case.  It would be closer to a generative use case it sounds like, which we haven't supported yet, but plan to explore this in future work.

Hacking the current code for generative tasks shouldn't be super difficult though, and we certainly welcome any contributions!

**Comment by thomasopsomer** [comment]
Just to follow on this topic, does the weights on HF also contain the `lm_head` weights ? And if not could you easily share those weights ? :)

**Comment by exnx** [comment]
yes the weights for the `lm_head` are there, but the provided loading ckpt code is implemented such that it replaces the `lm_head` with a new head (scratch) for a downstream task.  

If you literally will use the same pretraining model, ie no change to the architecture, then it's even easier, you just pass in the model ckpts using the flag `train.pretrained_model_path=path_to_ckpt`.  This will resume training from the ckpt - you would need to modify code to do inference on some dataset yourself.

Note: this doesn't just give you a DNA generator out of the box.  Remember, you're only predicting a single next token given a sequence.  If you want to generate a whole new sequence, then you'd have to do this in a loop and prepend 1 token at a time as you decode, which will be relatively slow.  There's a whole field dedicated to fast inference and decoding/generating (but we don't support on this repo).

Going to take some hacking :)


You would need to just modify that code.

**Comment by thomasopsomer** [comment]
Alright it's perfect, thanks for clearing that up :)

**Comment by exnx** [comment]
A way to get logits:
https://github.com/HazyResearch/hyena-dna#getting-logits-from-pretrained-model

To clarify though, it's not generating new sequences necessarily from this, it's generating 1 new token at the very end, so you'd to do a for loop to actually generate.

**Comment by gonzalobenegas** [comment]
Hi Eric, thank you for the update!

I've been playing since yesterday, after I saw announcements about implementation in HF.

I have a [simple Colab notebook](https://colab.research.google.com/drive/1Y6YOKuKRdvdkiU5jAdskPQ8B0xeGqeQn?usp=sharing) to evaluate the (averaged) log-likelihood of a sequence. Does it look correct?

BTW, is it possible there's a small bug [when computing the loss](https://huggingface.co/LongSafari/hyenadna-tiny-1k-seqlen-hf/blob/27d2b1c7c1373dd194f6f89fb44889bd67ce2ce9/modeling_hyena.py#L466)? I wonder if it should use `self.vocab_size`.

**Comment by Rocketknight1** [comment]
Hi @gonzalobenegas, you're totally right about the bug - my fault, I wrote that part of the port! It should be fixed now.

**Comment by exnx** [comment]
@gonzalobenegas looks good to me!

**Comment by gonzalobenegas** [comment]
Great, thank you!
